File: Homework/quoc_khanh/lesson_2/148.SortList.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def findMiddleNode(head: Optional[ListNode]):
    if head is None or head.next is None:
        return head
    fast, slow, prevSlow = head, head, None
    while fast != None and fast.next != None:
        prevSlow = slow
        slow = slow.next
        fast = fast.next
        if fast:
            fast = fast.next
    return prevSlow, slow

def mergeTwoSortedLinkedList(list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
    if list1 == None and list2 is None:
        return None
    if list1 == None:
        return list2
    if list2 == None:
        return list1
    pointer1, pointer2 = list1, list2
    
    dummyNode = ListNode()
    newTail = dummyNode
    next = None
    while pointer1 != None and pointer2 != None:
        if pointer1.val < pointer2.val:
            next = pointer1.next
            pointer1.next = None
            newTail.next = pointer1
            newTail = newTail.next
            pointer1 = next
        else:
            next = pointer2.next
            pointer2.next = None
            newTail.next = pointer2
            newTail = newTail.next
            pointer2 = next

    if pointer1 != None:
        newTail.next = pointer1
    if pointer2 != None:
        newTail.next = pointer2
    return dummyNode.next
             

class Solution:
    def sortList(self, node: Optional[ListNode]) -> Optional[ListNode]:
        if node == None or node.next == None:
            return node
        prevMiddle, middle = findMiddleNode(node)
        if prevMiddle:
            prevMiddle.next = None
        left = node
        list1 = self.sortList(left)
        list2 = self.sortList(middle)
        return mergeTwoSortedLinkedList(list1, list2)

File: Homework/Hoang_Nguyen/lesson3/148_sortList.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */

function middleNode(head) {
  let slow = head;
  let fast = head;
  while (slow.next && fast.next && fast.next.next) {
    slow = slow.next;
    fast = fast.next.next;
  }
  return slow;
}

function mergeLists(head1, head2) {
  let cur1 = head1;
  let cur2 = head2;
  let headResult = null;
  let curResult = null;
  while (cur1 && cur2) {
    let next;
    if (cur1.val < cur2.val) {
      next = cur1;
      cur1 = cur1.next;
    } else {
      next = cur2;
      cur2 = cur2.next;
    }
    if (curResult) {
      curResult.next = next;
    }
    curResult = next;

    if (!headResult) {
      headResult = curResult;
    }
  }
  if (cur1) {
    curResult.next = cur1;
  }
  if (cur2) {
    curResult.next = cur2;
  }
  return headResult;
}

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var sortList = function (head) {
  if (!head || !head.next) {
    return head;
  }
  // split list in halves
  let middle = middleNode(head);
  let rightHead = middle.next;
  middle.next = null;

  head = sortList(head);
  rightHead = sortList(rightHead);

  return mergeLists(head, rightHead);
};
